### PR TITLE
116 Generate Zoom links

### PR DESCRIPTION
**Pull request by ayomanetsan**
This PR adds a feature to create Zoom meeting links for the lesson. Upon the creation of a lesson, **two** links will be generated: one for the **host** and one for the **students**.

### CODE REVIEW

File: backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/ZoomApiService.cs [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+        }
+
+        using var httpClient = new HttpClient();
+        httpClient.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue(
+            "Bearer",
+            GenerateAccessToken()
+        );
+
+        var response = await httpClient.PostAsync(
+            "https://api.zoom.us/v2/users/me/meetings",
```
Comment (tatianahutii) [inline backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/ZoomApiService.cs:44]: it would be better to move this to the appsettings

Comment (tatianahutii) [reply backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/ZoomApiService.cs:44]: at least base url

File: backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/ZoomApiService.cs [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+                {
+                    topic = lessonTopic,
+                }),
+                Encoding.UTF8,
+                "application/json"
+            )
+        );
+
+        var responseBody = await response.Content.ReadAsStringAsync();
+        dynamic meeting = JsonConvert.DeserializeObject(responseBody);
```
Comment (tatianahutii) [inline backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/ZoomApiService.cs:56]: github actions warnings

**Review by tatianahutii — CHANGES_REQUESTED** [review]

File: backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/ZoomApiService.cs (hunk 2 above)
Comment (tatianahutii) [reply backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/ZoomApiService.cs:56]: and code smells

File: backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/ZoomApiService.cs [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+        var encodedHeader = Convert.ToBase64String(Encoding.UTF8.GetBytes(JsonConvert.SerializeObject(header)));
+        var encodedPayload = Convert.ToBase64String(Encoding.UTF8.GetBytes(JsonConvert.SerializeObject(payload)));
+
+        var signature = new HMACSHA256(Encoding.UTF8.GetBytes(_apiSecret))
+            .ComputeHash(Encoding.UTF8.GetBytes($"{encodedHeader}.{encodedPayload}"));
+
+        var encodedSignature = Convert.ToBase64String(signature);
+
+        return $"{encodedHeader}.{encodedPayload}.{encodedSignature}";
+    }
```
Comment (Anatolii-Ivanov) [inline backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/ZoomApiService.cs:87]: Can you move it to some sort of helper function?

File: backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/ZoomApiService.cs [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+    public async Task<ZoomMeetingLinks> GetMeetingLinks(string lessonTopic)
+    {
+        if (_token == null || DateTime.UtcNow > _tokenExpiration)
+        {
+            _token = GenerateAccessToken();
+            _tokenExpiration = DateTime.UtcNow.AddMinutes(90);
+        }
+
+        using var httpClient = new HttpClient();
```
Comment (Anatolii-Ivanov) [inline backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/ZoomApiService.cs:37]: Use IHttpClientFactory instead (as DI)

File: backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/ZoomApiService.cs [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+        httpClient.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue(
+            "Bearer",
+            GenerateAccessToken()
+        );
```
Comment (Anatolii-Ivanov) [inline backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/ZoomApiService.cs:41]: I'd move token generation and adding it to client to a separate function

File: backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/ZoomApiService.cs [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+        var response = await httpClient.PostAsync(
+            "https://api.zoom.us/v2/users/me/meetings",
+            new StringContent(
+                JsonConvert.SerializeObject(new
+                {
+                    topic = lessonTopic,
+                }),
+                Encoding.UTF8,
+                "application/json"
+            )
```
Comment (Anatolii-Ivanov) [inline backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/ZoomApiService.cs:52]: The same, to helper/private/static function (select one)

File: backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/ZoomApiService.cs (hunk 2 above)
Comment (Anatolii-Ivanov) [inline backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/ZoomApiService.cs:56]: Create a model with needed fields (it won't deserialize all other)

File: backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/ZoomApiService.cs [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+        );
+
+        var responseBody = await response.Content.ReadAsStringAsync();
+        dynamic meeting = JsonConvert.DeserializeObject(responseBody);
+
+        string joinUrl = meeting.join_url;
+        string startUrl = meeting.start_url;
+
+        return new ZoomMeetingLinks(joinUrl, startUrl);
+    }
```
Comment (Anatolii-Ivanov) [inline backend/EasySpeak.Core/EasySpeak.Core.BLL/Services/ZoomApiService.cs:62]: Can be done in one line

**Review by Anatolii-Ivanov — CHANGES_REQUESTED** [review]

File: backend/EasySpeak.Domain/EasySpeak.Domain.csproj
Comment (Anatolii-Ivanov) [inline backend/EasySpeak.Domain/EasySpeak.Domain.csproj:1]: Why you created this project?
Domain means main project with logic... In our case it's Core

**Pull request merged**